### ISSUE TITLE
Docker Compose

### ISSUE THREAD

**Issue by r3l1990** [rigon/photo-gallery#20]
Love the look and idea behind this.

Is there a docker compose file available for this?

**Comment by rigon** [comment]
Thank you so much!

There wasn't a docker compose file, but I was able to quickly write one, it is actually pretty simple: [checkout the example](https://github.com/rigon/photo-gallery#docker).

If you could provide any additional feedback on the project, I would really appreciate! 😃

**Comment by r3l1990** [comment]
Thank you so much - A lot of this is a little over my head but I will spin up an instance of this when home after work and let you know how I get on and report back if any issues.

**Comment by rigon** [comment]
Great, let me know if you need any help with it!

**Comment by r3l1990** [comment]
Span it up and it is working fine in most regards, I honestly love the fact it can play video that is not in h.264 format.

My folder structure is Year-Month-Day so most of my folders dont show any photos when I select them in the left hand pane, Is this normal behaviour?

I do have a couple of folders with just dumped photos and they display fine.

Is there a workaround to keep my file structure in place but be able to select a folder within a folder to view its contents or have I mucked something / not done something correctly to set it up?

Thanks Again

**Comment by r3l1990** [comment]
Just as an added thought maybe a button to switch beetwen show folders or show all in Year folder so you can switch photo views depending on what you are trying to do, find something specific from a certain date or just browse.

**Comment by rigon** [comment]
To play video, I just feed the browser straight up from the file, no transcoding in between. Being able to play it is up to the browser if supports the format. I have plans to fill the cases where the browser doesn't support, but it's a complex task and I'm leaving it for later.

**Comment by r3l1990** [comment]
Oh no the video side of things is perfect :) No issue at all its what I love about this.

My issue is I cannot actually see any of my photos that are in for example 2023's folder listed by month and then by date :)

My folder structure is Year-Month-Day so most of my folders dont show any photos when I select them in the left hand pane, Is this normal behaviour?

I do have a couple of folders with just dumped photos and they display fine.

Is there a workaround to keep my file structure in place but be able to select a folder within a folder to view its contents or have I mucked something / not done something correctly to set it up?

**Comment by rigon** [comment]
Regarding the folder structure, my ultimate goal is not forcing anyone in particular. Everyone has different opinions in how to organize their photos and I should respect that. However I started the project with the following assumption:

1. only scan for folders (i.e. the albums) in the collection root
2. only scan for photos the in the root of albums

Everything else that doesn't follow this structure doesn't show. It is a quite limiting and goes against the goals of the project, but it was working for me (most of the cases). And don't want you changing your stuff because this.

Recognizing this issue, I have been working lately on that:  #17
It is already merged on master, but not yet released. For the release want to test a bit more and add a few more things.
You can give it a try now by using this docker image: `rigon/photo-gallery:master`

A few things to have in mind (and I haven't yet written anywhere else):
- this project works best when you have a lot of albums
- not so much if you a have a lot of photos per album, as it attempts to load all photos within at once
- still doesn't perform too bad, I have an album with 20k+ photos and only takes 15s to load
- I have plans to address this issue by loading the album in chunks and virtual scrolling

You can give it a try by using this docker image: `rigon/photo-gallery:master`

**Comment by r3l1990** [comment]
Thank you I will give this a try just now - I can just add ":Master" to the end of rigon/photo-gallery in the docker compose yes?

If above works I will let you know how I get on.

Thank You :)

**Comment by r3l1990** [comment]
So when trying to pull the master image, it seems to be automatically exiting itself and not starting.

Trying to use it with Docker Compose like so:

version: "3"

volumes:
  photo-gallery_data:

services:
  photo-gallery:
    image: rigon/photo-gallery:master
    volumes:
      - photo-gallery_data:/thumbs
      - /media/data/photos/:/photos/:ro
      - /media/data/recent/:/recent/:rw
    ports:
      - 3080:3080
    command:
      - "-cname=Photos,path=/photos,thumbs=/thumbs"
      - "-cname=Recent,path=/recent,thumbs=/thumbs"

**Comment by rigon** [comment]
The docker compose file looks fine.
Maybe is because or the image or the container already exists and doesn't create a new one. Try delete them before.
If you could provide more info with logs from the console, I can help you further.

**Comment by r3l1990** [comment]
OK I have found the below log in portainer - Not sure if this is the one that will help you or not:

2023/06/21 16:22:14 Collections: map[Photos:Photos (/photos) Recent:Recent (/recent)]
2023/06/21 16:22:14 Current DB version v5 is different than required v6
2023/06/21 16:22:14 Run command with option -r enabled to recreate cache DB

Thank you again.

**Comment by rigon** [comment]
Ah ok, that requires upgrading the cache DB. Change your docker compose to include the flag `-r` in your command:

```yml
    command:
      - "-r"
      - "-cname=Photos,path=/photos,thumbs=/thumbs"
      - "-cname=Recent,path=/recent,thumbs=/thumbs"
```

Checkout the [Usage](https://github.com/rigon/photo-gallery#usage) section for more info.

**Comment by r3l1990** [comment]
Thank you :)

**Comment by rigon** [comment]
Thanks a lot!😄

I find the option to select a specific folder cool too, only when you have many sub-folders it becomes messy...

Yeah, creating the thumbnails takes a long time, but once they are cached after the first time, it is pretty quick!
You can create the thumbnails upfront in background, when the server is initiated, by adding the flag `-b` in the command.

For my entire collection which is about 1TB, took more than 2 days to finish on my Raspberry PI 4.

**Comment by r3l1990** [comment]
Thanks I have just reset the stck with -b command, mines is running on a ubuntu VM, will let you know the size of my collection and how long it takes to complete once done.

Thanks again.

**Comment by r3l1990** [comment]
190gb collection roughly 32000 files and the scan has taken about 17 hours, very snappy at loading photos now everything is cached.

Still seems to struggle with some video but like you say this will be down to the browser.

So far so good! Enjoying it may just replace photoview.

**Comment by rigon** [comment]
I'm really glad you found it useful. I will keep working on it to make it even better 😄 

Yeah, I still haven't paid too much attention to videos, I'm focused on making the app more robust and working fast.

**Comment by r3l1990** [comment]
Its been good with no issues, this will be a popular one in future :)

Video support would top its favour I think but that could just be me being bias, a self hosted alternative with transcoding simialr to Plex or JF is a dream :D

**Comment by rigon** [comment]
I hope you're right!
Just released v1.4.0, you can switch back to `latest` docker image, i.e. remove `:master`.
Have fun 😜
